File: ios/SignalKit/Sources/SignalKit/Queue/SignalStore.swift
import Foundation
import GRDB

public enum PendingSignalStatus: String, Codable, DatabaseValueConvertible {
    case pending
    case uploading
    case uploaded
    case failed
}

/// The persisted row shape, mirroring the upload envelope so a batch can be
/// built directly from a page of rows without further transformation.
struct PendingSignalRow: Codable, FetchableRecord, PersistableRecord {
    static let databaseTableName = "pending_signals"

    var id: String
    var deviceID: String
    var capturedAt: Date
    var signalType: String
    var payloadJSON: Data
    var lat: Double?
    var lon: Double?
    var horizontalAccuracy: Double?
    var locationAgeSeconds: Double?
    var status: PendingSignalStatus
    var batchID: String?
    var attemptCount: Int
    var createdAt: Date

    enum CodingKeys: String, CodingKey {
        case id, deviceID = "device_id", capturedAt = "captured_at", signalType = "signal_type"
        case payloadJSON = "payload_json", lat, lon
        case horizontalAccuracy = "horizontal_accuracy", locationAgeSeconds = "location_age_seconds"
        case status, batchID = "batch_id", attemptCount = "attempt_count", createdAt = "created_at"
    }
}

/// Offline-first queue for captured signals, backed by SQLite via GRDB.
/// Chosen over Core Data (too much ceremony for a flat high-frequency queue
/// table) and a raw file-based queue (weak status tracking, no atomic
/// partial-batch acknowledgement). A single `pending_signals` table with a
/// `status` column drives the whole pending -> uploading -> uploaded/failed
/// lifecycle that `SyncEngine` depends on.
public final class SignalStore {
    private let dbQueue: DatabaseQueue

    public init(path: String) throws {
        dbQueue = try DatabaseQueue(path: path)
        try migrator.migrate(dbQueue)
    }

    private var migrator: DatabaseMigrator {
        var migrator = DatabaseMigrator()
        migrator.registerMigration("createPendingSignals") { db in
            try db.create(table: "pending_signals") { t in
                t.column("id", .text).primaryKey()
                t.column("device_id", .text).notNull()
                t.column("captured_at", .datetime).notNull()
                t.column("signal_type", .text).notNull()
                t.column("payload_json", .blob).notNull()
                t.column("lat", .double)
                t.column("lon", .double)
                t.column("horizontal_accuracy", .double)
                t.column("location_age_seconds", .double)
                t.column("status", .text).notNull().defaults(to: PendingSignalStatus.pending.rawValue)
                t.column("batch_id", .text)
                t.column("attempt_count", .integer).notNull().defaults(to: 0)
                t.column("created_at", .datetime).notNull()
            }
            try db.create(index: "idx_pending_signals_status_captured", on: "pending_signals", columns: ["status", "captured_at"])
        }
        return migrator
    }

    public func enqueue(_ record: SignalRecord) throws {
        let row = try PendingSignalRow(
            id: record.id.uuidString,
            deviceID: record.deviceID.uuidString,
            capturedAt: record.capturedAt,
            signalType: record.payload.signalType.rawValue,
            payloadJSON: record.payload.encodedData(),
            lat: record.location?.lat,
            lon: record.location?.lon,
            horizontalAccuracy: record.location?.horizontalAccuracyMeters,
            locationAgeSeconds: record.location?.ageSeconds,
            status: .pending,
            batchID: nil,
            attemptCount: 0,
            createdAt: Date()
        )
        try dbQueue.write { db in
            try row.insert(db)
        }
    }

    /// Claims up to `limit` pending rows for upload, stamping them with
    /// `batchID` and moving them to `.uploading` so a crash mid-upload
    /// doesn't leave them silently stuck in `.pending` forever -- callers
    /// should call `markFailed` on timeout/crash recovery to requeue them.
    func claimBatch(batchID: String, limit: Int = 500) throws -> [PendingSignalRow] {
        try dbQueue.write { db in
            let rows = try PendingSignalRow
                .filter(Column("status") == PendingSignalStatus.pending.rawValue)
                .order(Column("captured_at"))
                .limit(limit)
                .fetchAll(db)

            for var row in rows {
                row.status = .uploading
                row.batchID = batchID
                try row.update(db)
            }
            return rows
        }
    }

    func markUploaded(ids: [String]) throws {
        guard !ids.isEmpty else { return }
        try dbQueue.write { db in
            try db.execute(
                sql: "UPDATE pending_signals SET status = ? WHERE id IN (\(ids.map { _ in "?" }.joined(separator: ","))) ",
                arguments: StatementArguments([PendingSignalStatus.uploaded.rawValue] + ids)
            )
        }
    }

    /// Reverts rejected/failed rows to `.pending` (transient) or `.failed`
    /// (permanent, after `maxAttempts`) so the sync engine's next pass
    /// picks up the right ones.
    func markFailed(ids: [String], permanent: Bool, maxAttempts: Int = 20) throws {
        guard !ids.isEmpty else { return }
        try dbQueue.write { db in
            for id in ids {
                guard var row = try PendingSignalRow.fetchOne(db, key: id) else { continue }
                row.attemptCount += 1
                row.status = (permanent || row.attemptCount >= maxAttempts) ? .failed : .pending
                row.batchID = nil
                try row.update(db)
            }
        }
    }

    /// Hard-deletes `.uploaded` rows older than `olderThan`, bounding local
    /// DB growth independent of upload success rate.
    public func pruneUploaded(olderThan cutoff: Date) throws {
        try dbQueue.write { db in
            try db.execute(
                sql: "DELETE FROM pending_signals WHERE status = ? AND captured_at < ?",
                arguments: [PendingSignalStatus.uploaded.rawValue, cutoff]
            )
        }
    }

    public func pendingCount() throws -> Int {
        try dbQueue.read { db in
            try PendingSignalRow.filter(Column("status") == PendingSignalStatus.pending.rawValue).fetchCount(db)
        }
    }
}

File: android/app/src/main/java/com/naberconsulting/signals/queue/SignalStore.kt
package com.naberconsulting.signals.queue

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import com.naberconsulting.signals.model.SignalRecord

/** A persisted queue row, mirroring the upload envelope. */
data class PendingSignalRow(
    val id: String,
    val capturedAtMillis: Long,
    val signalType: String,
    val payloadJson: String,
    val lat: Double?,
    val lon: Double?,
    val horizontalAccuracy: Double?,
    val locationAgeSeconds: Double?,
)

/**
 * Offline-first queue for captured signals, backed by SQLite. A single
 * `pending_signals` table with a `status` column drives the whole
 * pending -> uploading -> uploaded/failed lifecycle that [SyncEngine] depends
 * on -- the Android counterpart to the iOS GRDB-backed SignalStore.
 */
class SignalStore(context: Context) :
    SQLiteOpenHelper(context.applicationContext, DB_NAME, null, DB_VERSION) {

    init {
        // Rows left in `uploading` by a run that was killed mid-upload have no
        // in-flight tracking once the process exits -- requeue them so they
        // aren't silently stuck forever.
        recoverStaleUploads()
    }

    override fun onCreate(db: SQLiteDatabase) {
        db.execSQL(
            """
            CREATE TABLE pending_signals (
                id TEXT PRIMARY KEY NOT NULL,
                captured_at INTEGER NOT NULL,
                signal_type TEXT NOT NULL,
                payload_json TEXT NOT NULL,
                lat REAL,
                lon REAL,
                horizontal_accuracy REAL,
                location_age_seconds REAL,
                status TEXT NOT NULL DEFAULT 'pending',
                batch_id TEXT,
                attempt_count INTEGER NOT NULL DEFAULT 0,
                created_at INTEGER NOT NULL
            )
            """.trimIndent()
        )
        db.execSQL("CREATE INDEX idx_pending_status_captured ON pending_signals (status, captured_at)")
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        // Queue rows are ephemeral (pruned within 48h); a schema change can
        // safely drop and recreate rather than migrate.
        db.execSQL("DROP TABLE IF EXISTS pending_signals")
        onCreate(db)
    }

    private fun recoverStaleUploads() {
        writableDatabase.execSQL(
            "UPDATE pending_signals SET status = 'pending', batch_id = NULL WHERE status = 'uploading'"
        )
    }

    fun enqueue(record: SignalRecord) {
        val values = ContentValues().apply {
            put("id", record.id)
            put("captured_at", record.capturedAtMillis)
            put("signal_type", record.signalType.wire)
            put("payload_json", record.payload.toString())
            record.location?.let {
                put("lat", it.lat)
                put("lon", it.lon)
                it.horizontalAccuracyMeters?.let { v -> put("horizontal_accuracy", v) }
                it.ageSeconds?.let { v -> put("location_age_seconds", v) }
            }
            put("status", "pending")
            put("created_at", System.currentTimeMillis())
        }
        writableDatabase.insertWithOnConflict(
            "pending_signals", null, values, SQLiteDatabase.CONFLICT_IGNORE
        )
    }

    /**
     * Claims up to [limit] pending rows, stamping them with [batchId] and
     * moving them to `uploading` in one transaction so a crash mid-upload
     * doesn't leave them stuck in `pending`.
     */
    fun claimBatch(batchId: String, limit: Int = 500): List<PendingSignalRow> {
        val db = writableDatabase
        val rows = mutableListOf<PendingSignalRow>()
        db.beginTransaction()
        try {
            db.query(
                "pending_signals", null, "status = 'pending'", null,
                null, null, "captured_at ASC", limit.toString()
            ).use { c ->
                while (c.moveToNext()) rows.add(c.toRow())
            }
            if (rows.isNotEmpty()) {
                val ids = rows.joinToString(",") { "'${it.id}'" }
                val values = ContentValues().apply {
                    put("status", "uploading")
                    put("batch_id", batchId)
                }
                db.update("pending_signals", values, "id IN ($ids)", null)
            }
            db.setTransactionSuccessful()
        } finally {
            db.endTransaction()
        }
        return rows
    }

    fun markUploaded(ids: List<String>) {
        if (ids.isEmpty()) return
        val placeholders = ids.joinToString(",") { "?" }
        writableDatabase.execSQL(
            "UPDATE pending_signals SET status = 'uploaded' WHERE id IN ($placeholders)",
            ids.toTypedArray()
        )
    }

    /**
     * Reverts rejected/failed rows to `pending` (transient) or `failed`
     * (permanent, or after [maxAttempts]) so the next sync pass picks up the
     * right ones.
     */
    fun markFailed(ids: List<String>, permanent: Boolean, maxAttempts: Int = 20) {
        if (ids.isEmpty()) return
        val db = writableDatabase
        db.beginTransaction()
        try {
            for (id in ids) {
                db.query(
                    "pending_signals", arrayOf("attempt_count"), "id = ?", arrayOf(id),
                    null, null, null
                ).use { c ->
                    if (!c.moveToFirst()) return@use
                    val attempts = c.getInt(0) + 1
                    val status = if (permanent || attempts >= maxAttempts) "failed" else "pending"
                    val values = ContentValues().apply {
                        put("attempt_count", attempts)
                        put("status", status)
                        putNull("batch_id")
                    }
                    db.update("pending_signals", values, "id = ?", arrayOf(id))
                }
            }
            db.setTransactionSuccessful()
        } finally {
            db.endTransaction()
        }
    }

    /** Hard-deletes `uploaded` rows older than [cutoffMillis], bounding DB growth. */
    fun pruneUploaded(cutoffMillis: Long) {
        writableDatabase.execSQL(
            "DELETE FROM pending_signals WHERE status = 'uploaded' AND captured_at < ?",
            arrayOf<Any>(cutoffMillis)
        )
    }

    fun pendingCount(): Int =
        readableDatabase.rawQuery(
            "SELECT COUNT(*) FROM pending_signals WHERE status = 'pending'", null
        ).use { c -> if (c.moveToFirst()) c.getInt(0) else 0 }

    private fun android.database.Cursor.toRow(): PendingSignalRow {
        fun optDouble(name: String): Double? {
            val i = getColumnIndexOrThrow(name)
            return if (isNull(i)) null else getDouble(i)
        }
        return PendingSignalRow(
            id = getString(getColumnIndexOrThrow("id")),
            capturedAtMillis = getLong(getColumnIndexOrThrow("captured_at")),
            signalType = getString(getColumnIndexOrThrow("signal_type")),
            payloadJson = getString(getColumnIndexOrThrow("payload_json")),
            lat = optDouble("lat"),
            lon = optDouble("lon"),
            horizontalAccuracy = optDouble("horizontal_accuracy"),
            locationAgeSeconds = optDouble("location_age_seconds"),
        )
    }

    companion object {
        private const val DB_NAME = "signals.db"
        private const val DB_VERSION = 1
    }
}
